cmd/log/src/lib.rs

```diff
@@ -0,0 +1,288 @@
+//! the argument as a substring will be displayed.  
+//!
+
+use anyhow::{bail, Result};
+use clap::Command as ClapCommand;
+use clap::{CommandFactory, Parser};
+use humility::cli::Subcommand;
+use humility::core::Core;
+use humility::hubris::*;
+use humility::reflect::{self, Load, Value};
+use humility_cmd::doppel::{StaticCell, Stringbuf};
+use humility_cmd::{Archive, Attach, Command, Validate};
+use std::str;
+use std::thread;
+use std::time;
+
+#[derive(Parser, Debug)]
+#[clap(name = "stringbuf", about = env!("CARGO_PKG_DESCRIPTION"))]
+struct StringbufArgs {
+    /// list all stringbufs
+    #[clap(long, short)]
+    list: bool,
+    /// continously update the log
+    #[clap(long, short, conflicts_with = "list")]
+    monitor: bool,
+    /// how long to delay in ms between polling for new log entries
+    #[clap(
+        long,
+        short,
+        default_value = "100",
+        requires = "monitor",
+        conflicts_with = "list"
+    )]
+    delay: u64,
+    /// print only a single stringbuffer by substring of name
+    #[clap(conflicts_with = "list")]
+    name: Option<String>,
+}
+
+fn load_stringbuf(
+    hubris: &HubrisArchive,
+    core: &mut dyn Core,
+    definition: &HubrisStruct,
+    ringbuf_var: &HubrisVariable,
+) -> Stringbuf {
+    let mut buf: Vec<u8> = vec![];
+    buf.resize_with(ringbuf_var.size, Default::default);
+    // Load the stringbuf data from buf
+    core.op_start().unwrap();
+    core.read_8(ringbuf_var.addr, buf.as_mut_slice()).unwrap();
+    core.op_done().unwrap();
+
+    // use the buf to create a stringbuf struct
+    let ringbuf_val: Value = Value::Struct(
+        reflect::load_struct(hubris, buf.as_mut_slice(), definition, 0)
+            .unwrap(),
+    );
+    let cell: StaticCell = StaticCell::from_value(&ringbuf_val).unwrap();
+    Stringbuf::from_value(&cell.cell.value).unwrap()
+}
+
+fn stringbuf_read(
+    hubris: &HubrisArchive,
+    core: &mut dyn Core,
+    definition: &HubrisStruct,
+    ringbuf_var: &HubrisVariable,
+    prev_last_idx: u64,
+) -> Result<(String, u64)> {
+    let mut log_msg: String = "".to_owned();
+
+    // load the stringbuf from hubris into a corresponding local struct
+    let stringbuf: Stringbuf =
+        load_stringbuf(hubris, core, definition, ringbuf_var);
+
+    // extract the log itself as a [u8]
+    let buffer = stringbuf.buffer.as_slice();
+    // the last written element in the log
+    let last: u64 = stringbuf.last.unwrap() as u64;
+
+    // read everything between current last, and last last by
+    // generating index range which connects these two in a circular manner, only including the
+    // new characters, and in the order they should be printed
+    // prev_last_idx has been printed already, last has not
+    let new_idxs: Vec<u64> = if last > prev_last_idx {
+        (prev_last_idx + 1..last).collect()
+    } else {
+        // This logic also works for last == prev_last_idx since we just need to print the entire range
```

Comment:
Does this result in garbage bytes being added from prev_last_idx..buffer_len in that case?